File: app/src/main/java/com/gmail/cristiandeives/motofretado/TrackBusMvp.java
package com.gmail.cristiandeives.motofretado;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.UiThread;

import com.gmail.cristiandeives.motofretado.http.Bus;
import com.gmail.cristiandeives.motofretado.http.ModelListener;

import java.util.List;

interface TrackBusMvp {
    interface Model {
        @UiThread
        void readAllBuses(@NonNull ModelListener<Bus[]> listener);
        @UiThread
        void createBus(@NonNull Bus bus, ModelListener<Bus> listener);
        @UiThread
        void cancelAllRequests();
    }

    interface View {
        @UiThread
        @Nullable String getBusId();
        @UiThread
        void displayMessage(@Nullable String text);
        @UiThread
        void enableBusId();
        @UiThread
        void disableBusId();
        @UiThread
        void uncheckSwitchDetectAutomatically();
        @UiThread
        void setAvailableBuses(@NonNull List<Bus> buses, @Nullable String selectedBusId);
    }

    interface Presenter {
        @UiThread
        void onAttach(View view);
        @UiThread
        void onDetach();
        @UiThread
        void startLocationUpdate();
        @UiThread
        void stopLocationUpdate();
        @UiThread
        void startActivityDetection();
        @UiThread
        void stopActivityDetection();
        @UiThread
        void createBus(@NonNull Bus bus);
    }
}

File: app/src/main/java/com/gmail/cristiandeives/motofretado/TrackBusMvp.kt
package com.gmail.cristiandeives.motofretado

import android.support.annotation.UiThread
import com.gmail.cristiandeives.motofretado.http.Bus
import com.gmail.cristiandeives.motofretado.http.ModelListener

@UiThread
internal interface TrackBusMvp {
    interface Model {
        fun readAllBuses(listener: ModelListener<List<Bus>>)
        fun createBus(bus: Bus, listener: ModelListener<Bus>)
        fun cancelAllRequests()
    }

    interface View {
        fun getBusId(): String?
        fun displayMessage(text: String)
        fun enableBusId()
        fun disableBusId()
        fun uncheckSwitchDetectAutomatically()
        fun setAvailableBuses(buses: List<Bus>, selectedBusId: String?)
        fun setBusError(errorMessage: String)
   }

    interface Presenter {
        fun onAttach(view: View)
        fun onDetach()
        fun startLocationUpdate()
        fun stopLocationUpdate()
        fun startActivityDetection()
        fun stopActivityDetection()
        fun createBus(bus: Bus)
    }
}